File: mdbook/src/mapreduce/cpp/openacc/naive.cpp
#include <iostream>
#include <timer.h>

using namespace std;

int main() {
  int num = 100000000;
  int *elements = new int[num];
  long res = 0;

  timer time;

  #pragma acc parallel loop copy(elements[:num])
  for (int i = 0; i < num; i++) {
    elements[i] = i;
  }

  #pragma acc parallel loop reduction(+: res) copy(elements[:num])
  for (int i = 0; i < num; i++) {
    res += elements[i];
  }

  cout << "Elapsed time: " << time.getTime() << endl;

  cout << "The result is: " << res << endl;
}

File: mdbook/src/mapreduce/fortran/openacc/naive.f90
PROGRAM mapreduce
    integer, parameter :: num = 100000000
    integer*8 :: res = 0
    integer, dimension(:), allocatable :: elements
    real :: start_time, end_time

    allocate(elements(num))

    call cpu_time(start_time)
    !$acc parallel loop copyout(elements)
    do i=1,num
        elements(i) = i
    enddo

    !$acc parallel loop copyin(elements) reduction(+:res)
    do i=1,num
        res = res + elements(i)
    enddo
    call cpu_time(end_time)

    print *, "time:", end_time - start_time, "seconds"
    print *, "the result is:", res

END PROGRAM mapreduce
